src/main.rs

```diff
@@ -86,8 +89,23 @@ async fn provision() -> Result<(), anyhow::Error> {
         .default_headers(default_headers)
         .build()?;
 
-    let instance_metadata = imds::query(&client).await?;
-    let username = get_username(&instance_metadata, &get_environment()?)?;
+    // Username can be obtained either via fetching instance metadata from IMDS
+    // or mounting a local device for OVF environment file. It should not fail
+    // immediately in a single failure, instead it should fall back to the other
+    // mechanism. So it is not a good idea to use `?` for query() or
+    // get_environment().
+    let instance_metadata = match imds::query(&client).await {
+        Ok(m) => Some(m),
+        Err(_) => None,
+    };
```

Comment:
There's a [convenience function](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) on `Result` to convert to `Option` (and [one to go back the other way](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or)):
```suggestion
    let instance_metadata = match imds::query(&client).await.ok();
```

Reply:
Good catch, fixed.